main.c

```diff
@@ -146,18 +157,20 @@ static void linux_boot(const struct boot_params boot_params)
     die_on(LINUX_HEADER_OFFSET + linux_header_size > sizeof(*linux_params),
            "Invalid linux header size");
 
-    memcpy((char *)linux_params + LINUX_HEADER_OFFSET,
-           (const char *)(boot_params.kernel_addr + LINUX_HEADER_OFFSET),
+    memcpy((uint8_t *)linux_params + LINUX_HEADER_OFFSET,
+           (const uint8_t *)(actual_kernel_load_addr + LINUX_HEADER_OFFSET),
```

Comment:
The second parameter to `memcpy` is the source. You did not relocate anything yet, so there is probably garbage at this address.